pkg/goDB/conditions/node/filter_direction.go

```diff
@@ -0,0 +1,160 @@
+	// filters
+	errMultipleDirectionFilterConditions = errors.New("multiple direction filters specified")
+
+	// errMisplacedDirectionFilterCondition indicates that
+	// the specified condition contains a misplaced traffic direction
+	// filter
+	errMisplacedDirectionFilterCondition = errors.New("misplaced direction filter")
+
+	unsupportedDirectionFilterComparatorStr = "unsupported direction filter comparator: %s"
+	unsupportedDirectionFilterStr           = "unsupported direction filter: %s (use one of 'in', 'out', 'uni', 'bi')"
+)
+
+// isDirectionCondition returns an error if node represents a
+// direction filter condition, and the node itself otherwise.
+// It is used in combination with node.transform() in order
+// to ensure direction filter conditions occur only at the
+// top level of the node or the second level in case the
+// top level represents a conjunction (AND).
+func isDirectionCondition(node conditionNode) (Node, error) {
+	if node.attribute == types.FilterKeywordDirection {
+		return nil, errMisplacedDirectionFilterCondition
+	}
+	return node, nil
+}
+
+// extractDirectionFilter returns a ValFilter if node represents a valid direction filter condition.
+// If node represents an invalid direction filter (e.g., unsupported direction filter comparator),
+// nil is returned with an error describing the invalidity reason.
+// If the node does not contain a filter condition, nil is returned without any error.
+func extractDirectionFilter(node Node) (hashmap.ValFilter, error) {
+	switch node := node.(type) {
+	case conditionNode:
+		if node.attribute != types.FilterKeywordDirection && node.attribute != types.FilterKeywordDirectionSugared {
+			return nil, nil
+		}
+		if node.comparator != "=" {
+			return nil, fmt.Errorf(unsupportedDirectionFilterComparatorStr, node.comparator)
+		}
+		var filter hashmap.ValFilter
+		switch FilterTypeDirection(node.value) {
+		case filterTypeDirectionIn, filterTypeDirectionInSugared:
+			filter = types.Counters.IsOnlyInbound
+		case filterTypeDirectionOut, filterTypeDirectionOutSugared:
+			filter = types.Counters.IsOnlyOutbound
+		case filterTypeDirectionUni, filterTypeDirectionUniSugared:
+			filter = types.Counters.IsUnidirectional
+		case filterTypeDirectionBi, filterTypeDirectionBiSugared:
+			filter = types.Counters.IsBidirectional
+		default:
+			return nil, fmt.Errorf(unsupportedDirectionFilterStr, node.value)
+		}
+		return filter, nil
+	default:
+		return nil, nil
+	}
+}
+
+// splitOffDirectionFilter splits off the traffic direction filter
+// from the node, and returns the node without the direction filter,
+// and a separate ValFilterNode representing the direction filter.
+// The direction filter placement inside the node is being validated.
+// If there is no direction filter included in node, it returns the node
+// and a ValFilterNode with FilterType FilterKeywordNone.
+func splitOffDirectionFilter(node Node) (Node, ValFilterNode, error) {
```

Comment:
I'd be surprised if this one doesn't trigger the cyclomatic complexity Linter... :rofl: TBH I'm having trouble following the flow / logic here. Probably it's a total nitpick, but maybe it could be split / refactored in a way that makes it more readable?

Reply:
Thanks for pointing it out, I can see it now too :laughing:  
I am working on increasing the function's readability.

Reply:
Thanks for extending the comments on this one, surely helps following what's going on!